comp/networkpath/npscheduler/npschedulerimpl/npscheduler.go

```diff
@@ -7,31 +7,324 @@
 }
 
-func (s *npSchedulerImpl) ScheduleConns(_ []*model.Connection) {
+func newNoopNpSchedulerImpl() *npSchedulerImpl {
+	return &npSchedulerImpl{enabled: false}
+}
+
+func newNpSchedulerImpl(epForwarder eventplatform.Forwarder, logger log.Component, sysprobeYamlConfig config.Reader) *npSchedulerImpl {
+	workers := sysprobeYamlConfig.GetInt("network_path.workers")
+	pathtestInputChanSize := sysprobeYamlConfig.GetInt("network_path.input_chan_size")
+	pathtestProcessChanSize := sysprobeYamlConfig.GetInt("network_path.process_chan_size")
+	pathtestTTL := sysprobeYamlConfig.GetDuration("network_path.pathtest_ttl")
+	pathtestInterval := sysprobeYamlConfig.GetDuration("network_path.pathtest_interval")
+	flushInterval := sysprobeYamlConfig.GetDuration("network_path.flush_interval")
+
+	logger.Infof("New NpScheduler (workers=%d input_chan_size=%d pathtest_ttl=%s pathtest_interval=%s)",
+		workers,
+		pathtestInputChanSize,
+		pathtestTTL.String(),
+		pathtestInterval.String())
+
+	return &npSchedulerImpl{
+		enabled:     true,
+		epForwarder: epForwarder,
+		logger:      logger,
+
+		pathtestStore:       pathteststore.NewPathtestStore(pathtestTTL, pathtestInterval, logger),
+		pathtestInputChan:   make(chan *common.Pathtest, pathtestInputChanSize),
+		pathtestProcessChan: make(chan *pathteststore.PathtestContext, pathtestProcessChanSize),
+		flushInterval:       flushInterval,
+		workers:             workers,
+
+		metricSender: metricsender.NewMetricSenderStatsd(statsd.Client),
+
+		receivedPathtestCount:    atomic.NewUint64(0),
+		processedTracerouteCount: atomic.NewUint64(0),
+		TimeNowFn:                time.Now,
+
+		stopChan:      make(chan struct{}),
+		runDone:       make(chan struct{}),
+		flushLoopDone: make(chan struct{}),
+
+		statsdClient: statsd.Client,
+
+		runTraceroute: runTraceroute,
+	}
+}
+
+func (s *npSchedulerImpl) ScheduleConns(conns []*model.Connection) {
 	if !s.enabled {
 		return
 	}
-	// TODO: IMPLEMENTATION IN SEPARATE PR (to make PRs easier to review)
+	startTime := time.Now()
+	for _, conn := range conns {
+		// Only process outgoing traffic
+		if !shouldScheduleNetworkPathForConn(conn) {
+			continue
+		}
+		remoteAddr := conn.Raddr
+		remotePort := uint16(conn.Raddr.Port)
+		err := s.scheduleOne(remoteAddr.Ip, remotePort)
+		if err != nil {
+			s.logger.Errorf("Error scheduling pathtests: %s", err)
+		}
+	}
+
+	scheduleDuration := time.Since(startTime)
+	s.statsdClient.Gauge("datadog.network_path.scheduler.schedule_duration", scheduleDuration.Seconds(), nil, 1) //nolint:errcheck
 }
 
-func newNoopNpSchedulerImpl() *npSchedulerImpl {
-	return &npSchedulerImpl{
-		enabled: false,
+func shouldScheduleNetworkPathForConn(conn *model.Connection) bool {
+	return conn.Direction == model.ConnectionDirection_outgoing
+}
+
+// scheduleOne schedules pathtests.
+// It shouldn't block, if the input channel is full, an error is returned.
+func (s *npSchedulerImpl) scheduleOne(hostname string, port uint16) error {
```

Comment:
hostname -> ip would be more accurate?

Reply:
In case of NPM connections it's an IP Address, 

but I used hostname since `scheduleOne` could be used in the future for other cases where it might a more generic hostname (domain or IP).

Reply:
You do have a call to `net.ParseIP` below that will fail if `hostname` is not an IP.